File: frontend/src/services/monitoring.ts
import { api } from './api';

export interface UptimeStats {
  uptimePercent: number;
  totalChecks: number;
  failedChecks: number;
  avgResponseTime: number;
}

export interface MonitoringCheck {
  id: string;
  projectId: string;
  target: 'frontend' | 'backend';
  status: 'up' | 'degraded' | 'down' | 'unknown';
  statusCode: number | null;
  responseTime: number | null;
  errorType: string | null;
  errorMessage: string | null;
  checkedAt: string;
}

export interface Incident {
  id: string;
  projectId: string;
  target: string;
  status: 'open' | 'resolved';
  startedAt: string;
  resolvedAt: string | null;
  failureCount: number;
  lastError: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface MonitoringUpdate {
  monitoringEnabled?: boolean;
  monitoringInterval?: number;
  frontendUrl?: string;
  backendUrl?: string;
}

export const monitoringService = {
  updateSettings: async (projectId: string, data: MonitoringUpdate): Promise<void> => {
    await api.patch(`/projects/${projectId}/monitoring`, data);
  },

  getHistory: async (projectId: string, limit: number = 50): Promise<MonitoringCheck[]> => {
    const response = await api.get(`/projects/${projectId}/monitoring/history`, { params: { limit } });
    return response.data;
  },

  getIncidents: async (projectId: string): Promise<Incident[]> => {
    const response = await api.get(`/projects/${projectId}/incidents`);
    return response.data;
  },

  getUptime: async (projectId: string, days: number = 1): Promise<UptimeStats> => {
    const response = await api.get(`/projects/${projectId}/uptime`, { params: { days } });
    return response.data;
  }
};

File: backend/app/schemas/monitoring.py
from pydantic import BaseModel, HttpUrl
from typing import Optional, List
from datetime import datetime
from enum import Enum

class TargetType(str, Enum):
    FRONTEND = "frontend"
    BACKEND = "backend"

class MonitoringStatus(str, Enum):
    UP = "up"
    DEGRADED = "degraded"
    DOWN = "down"
    UNKNOWN = "unknown"

class IncidentStatus(str, Enum):
    OPEN = "open"
    RESOLVED = "resolved"

class MonitoringCheck(BaseModel):
    id: str
    projectId: str
    target: TargetType
    status: MonitoringStatus
    statusCode: Optional[int] = None
    responseTime: Optional[int] = None # in ms
    errorType: Optional[str] = None
    errorMessage: Optional[str] = None
    checkedAt: datetime

    class Config:
        populate_by_name = True

class Incident(BaseModel):
    id: str
    projectId: str
    target: TargetType
    status: IncidentStatus
    startedAt: datetime
    resolvedAt: Optional[datetime] = None
    failureCount: int
    lastError: Optional[str] = None
    createdAt: datetime
    updatedAt: datetime

    class Config:
        populate_by_name = True

class UptimeStats(BaseModel):
    uptimePercent: float
    totalChecks: int
    failedChecks: int
    avgResponseTime: int # in ms
